### PR TITLE
feat: Add support for Terraform resource (aws_sns_topic_subscription)

### PR DESCRIPTION
**Pull request by dhoppe**
## Description
This pull request will add support for the Terraform resource:
* aws_sns_topic_subscription

## Motivation and Context
This Terraform module should be able to manage SNS topic subscriptions.

## Breaking Changes
This change does not break anything, because this feature is disabled by default.

## How Has This Been Tested?
- [x] I have tested these changes using `terraform plan` and `terraform apply`, locally.

### CODE REVIEW

File: variables.tf
Comment (dhoppe) [inline variables.tf:163]: There is a naming conflict and I am not sure if SNS topic and SNS topic subscriptions use different delivery policies.

**Review by antonbabenko — CHANGES_REQUESTED** [review]
Good start, let's extend this support from the very beginning.

1. There should be support for multiple subscriptions to one SNS topic. Use `for_each` instead of `count`.
2. This module should create SNS topic and attach aws_sns_topic_subscriptions to it. Make an argument `subscriptions` (`map(any)` or `any`) which will accept all arguments for individual subscriptions (see [eventbridge module](https://github.com/terraform-aws-modules/terraform-aws-eventbridge/#eventbridge-complete), for example).
3. Make changes to the code in `examples` folder and show all possible ways to subscribe (see `eventbridge` module or pretty much any other module in `terraform-aws-modules`, for eg). That code must work right away (without any changes from the user).

**Comment by dhoppe** [comment]
@antonbabenko This is harder than I expected. 😄 

1. Is this really the scope of a Terraform child module? After all, `for_each` can be used inside a Terraform root module. 
```
locals {
  platforms = ["qa-1"]
  service = "my-service"
}

module "aws-sqs_dead-letter-queue" {
  source  = "terraform-aws-modules/sqs/aws"
  version = "~> 3.2.1"

  for_each = toset(local.platforms)

  name = "${each.value}-${local.sqs_queue}-dlq"

  tags = {
    Environment = var.environment
    Service     = local.service
  }
}

module "aws-sqs_queue" {
  source  = "terraform-aws-modules/sqs/aws"
  version = "~> 3.2.1"

  for_each = toset(local.platforms)

  name                       = "${each.value}-${local.sqs_queue}"
  visibility_timeout_seconds = 30
  message_retention_seconds  = 1209600
  receive_wait_time_seconds  = 20
  redrive_policy = jsonencode(
    {
      deadLetterTargetArn = module.aws-sqs_dead-letter-queue[each.key].sqs_queue_arn
      maxReceiveCount     = 5
    }
  )

  tags = {
    Environment = var.environment
    Service     = local.service
  }
}

module "aws-sns" {
  source  = "terraform-aws-modules/sns/aws"
  version = "~> 3.3.0"

  for_each = toset(local.platforms)

  endpoint             = module.aws-sqs_queue[each.key].sqs_queue_arn
  protocol             = "sqs"
  topic_arn            = data.sops_file.secrets.data["sns.topic_arn"]
  raw_message_delivery = true

  tags = {
    Environment = var.environment
    Service     = local.service
  }
}
```
I know that this creates multiple directories, but it also supports different versions of a Terraform child module or different providers (cross accounts).
```
➜ ls -l | egrep '(sns|sqs)'
drwxr-xr-x 16 dhoppe staff 512 Jan 26 14:06 aws-sns
drwxr-xr-x 16 dhoppe staff 512 Jan 26 14:06 aws-sqs_dead-letter-queue
drwxr-xr-x 10 dhoppe staff 320 Jan 26 14:06 aws-sqs_queue
```
2. I would always use a local variable at this point to pass different arguments to the Terraform child module. This will also support different endpoints or protocols. 
3. I did not adjust the examples because SNS topics are not even included in them. The topic was already mentioned in Issue #5. 

Users should always look at the documentation of AWS or Terraform. This way they can point out that this module does not have support for AWS Firehose.

To be honest, I just wanted to add the missing Terraform resource. I need to migrate all AWS environments of the company to Terraform Cloud and in this context I will definitely create additional pull requests or provide new Terraform modules. 

Everything else I can do in my spare time. 😄

**Comment by dhoppe** [comment]
@antonbabenko @barryib @bryantbiggs Anyone?

**Comment by antonbabenko** [comment]
Yes, it is a bit more work to get it merged into the repository. Not just from yourself but also for us (maintainers) since we need to maintain this functionality and integrations in the future. If there will be missing examples or incorrect implementation, it will be harder for us to do without introducing breaking changes into the module.

If you need to have this functionality yourself, please use your fork while you or someone else finish this PR and we can get it merged.

Sorry that it takes some time.

**Comment by dhoppe** [comment]
@antonbabenko I really do not get your point. 

* The module did not have a proper example before. 
* The new resource is disabled by default.
* The README.md contains a example and the new variables.

The only thing that might be different is the `protocol`, but the purpose of this module is not to explain a user how AWS works. For anything else a user would look at the [docs](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/sns_topic_subscription) of the Terraform provider.

**Comment by antonbabenko** [comment]
When there is a feature in the module (not just this but pretty much [any other terraform-aws-modules](https://github.com/terraform-aws-modules/) ) we really strive to show it in a working example.

Without an example and the rest of things updated, we won't be merging this PR. It is faster and better that you use your own fork in the meanwhile.

**Comment by dhoppe** [comment]
@antonbabenko Done.

**Comment by dhoppe** [comment]
Still waiting for review / merge of this pull request.

**Comment by hieultan** [comment]
Hi there, any update? Still waiting for review

**Comment by antonbabenko** [comment]
Sorry but this PR still lacks a few important improvements and in a current way, I won't merge it. I described the required improvements in the comments above (primarily [this comment](https://github.com/terraform-aws-modules/terraform-aws-sns/pull/29#pullrequestreview-863493845)).

If you want to use the functionality as it is, please use the fork.

**Comment by hieultan** [comment]
Get your point, thanks

**Pull request closed without merging**
